### PR TITLE
[deckhouse-controller] refactor releases metrics

### LINKED ISSUE #9684: [deckhouse-controller] Fire an alert about blocked ModuleRelease
### Preflight Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/deckhouse/deckhouse/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://github.com/deckhouse/deckhouse/issues) for an issue that matches the one I want to file, without success.

### Use case. Why is this important?

ModuleRelease could be blocked and we don't known about it

### Proposed Solution

Add metrics for all extenders and fire an alert like [this](https://github.com/deckhouse/deckhouse/blob/main/modules/340-monitoring-deckhouse/monitoring/prometheus-rules/alerting.yaml#L73)

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by miklezzzz**
## Description
Refactors deckhouse and module controllers (in terms of processing pending releases).
Implements two new metrics `d8_release_info` and `d8_module_release_info` to reflect some of the statuses (manual approval required, requirements not met, disruption approval required, etc) releases can take on, and corresponding alerting.

## Why do we need it, and what problem does it solve?
Implements a slightly more straightforward way of providing release metrics (d8_release_info/d8_module_release_info) and corresponding alerting.
The following changes were introduced to the release processing logic:
- PhaseSuspended is avoided due to lack of options, from a user's point of view, what it's possible to do with a suspend release (it's neither deleted nor possible to 'revive' such a release);
- Transition time of a release is only changed if its phase is changed, changing transition time on every message update results in having an inconsistent and baffling timer; 
- Some strange logic of restoring current deployed release when processing pending releases was removed from the module release controller; releases are installed either via `updater` or when `restoreAbsentReleaseses` is called on start;
- a new message for pending releases is introduced - `Awaiting %s release to be deployed`, that is set to releases that can't be deployed due to another release awaiting to be deployed.
Closes #9684 

## What is the expected result?
Module and Deckhouse releases metrics are updated according to their states.

## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: deckhouse-controller
type: chore
summary: Refactor release processing.
impact_level: default
```

### CODE REVIEW

**Comment by miklezzzz** [comment]
a module is enabled and its release is blocked by requirements
image

a module is enabled and its release is waiting for manual approve
image

a deckhouse release is waiting for manual approve and there are 3 more releases waiting for being deployed
image

2 more releases
image

1 more release
image

**Comment by miklezzzz** [comment]
deckhouse release disruption approval required
image

deckhouse releases are waiting for another release to be deployed
```
root@dev-master-0:~# kubectl  get deckhousereleases.deckhouse.io 
NAME       PHASE        TRANSITIONTIME   MESSAGE
v1.63.10   Skipped      4d3h             
v1.63.11   Skipped      4d23h            
v1.63.14   Skipped      4d5h             
v1.63.15   Superseded   4d3h             
v1.63.16   Superseded   3d5h             
v1.63.17   Deployed     3d5h             
v1.64.11   Pending      27m              Release is waiting for the 'release.deckhouse.io/approved: "true"' annotation
v1.66      Pending      27m              Awaiting for v1.64.11 release to be deployed
v2.1.0     Pending      19m              Awaiting for v1.64.11 release to be deployed
```
image

deckhouse release is blocked by some module with deckhouseVersion requirements
```
kubectl  get deckhousereleases.deckhouse.io 
NAME       PHASE        TRANSITIONTIME   MESSAGE
v1.63.10   Skipped      4d3h             
v1.63.11   Skipped      4d23h            
v1.63.14   Skipped      4d5h             
v1.63.15   Superseded   4d3h             
v1.63.16   Superseded   3d5h             
v1.63.17   Deployed     3d5h             
v1.64.11   Pending      28m              Release is waiting for the 'release.deckhouse.io/approved: "true"' annotation
v1.66      Pending      28m              Awaiting for v1.64.11 release to be deployed
v2.1.0     Pending      21m              requirements of the 'echo' module are not satisfied: 2.1.0 deckhouse version is not suitable: 2.1.0 is greater than or equal to v2.1.0
```
image

deckhouse release requirements aren't met
```
kubectl  get deckhousereleases.deckhouse.io 
NAME       PHASE        TRANSITIONTIME   MESSAGE
v1.63.10   Skipped      4d3h             
v1.63.11   Skipped      4d23h            
v1.63.14   Skipped      4d5h             
v1.63.15   Superseded   4d3h             
v1.63.16   Superseded   3d5h             
v1.63.17   Deployed     3d5h             
v1.64.11   Pending      48m              Release is waiting for the 'release.deckhouse.io/approved: "true"' annotation
v1.66      Pending      48m              "ingressNginx" requirement for DeckhouseRelease "1.66.0" not met: minimal IngressNginxController version is lower then required
v2.1.0     Pending      41m              requirements of the 'echo' module are not satisfied: 2.1.0 deckhouse version is not suitable: 2.1.0 is greater than or equal to v2.1.0
```

no blocks
```
kubectl  get deckhousereleases.deckhouse.io 
NAME       PHASE        TRANSITIONTIME   MESSAGE
v1.63.10   Skipped      4d3h             
v1.63.11   Skipped      4d23h            
v1.63.14   Skipped      4d5h             
v1.63.15   Superseded   4d3h             
v1.63.16   Superseded   3d5h             
v1.63.17   Deployed     3d5h             
v1.64.11   Pending      51m              Release is waiting for the update window until 01 Nov 24 07:30 UTC
v1.66      Pending      51m              Awaiting for v1.64.11 release to be deployed
v2.1.0     Pending      44m              Awaiting for v1.64.11 release to be deployed
```

**Comment by miklezzzz** [comment]
module release has unsatisfied requirements
```
root@dev-master-0:~# kubectl  get mr | grep echo
echo-v0.5.2                              Pending    losev-echo      24s              requirements are not satisfied: current kubernetes version is not suitable: 1.29.9 is less than or equal to v1.35.0
```
image


module release is waiting for manual approval
```
root@dev-master-0:~# kubectl  get mr | grep echo
echo-v0.5.2                              Pending    losev-echo      110s             Release is waiting for the 'modules.deckhouse.io/approved: "true"' annotation
```
image


module can't start as its requirements aren't satisfied
```
root@dev-master-0:~# kubectl  get module echo -o wide
NAME   WEIGHT   STATE      SOURCE       STAGE   STATUS   MESSAGE
echo   900      Disabled   losev-test                    Info: turned off by KubernetesVersion constraint, requirements are not satisfied: current kubernetes version is not suitable: 1.29.9 is less than or equal to v1.35.0
```

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @miklezzzz. I've started the workflow for label 'e2e/run/gcp'...

**Review by z9r5 — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @miklezzzz. I've started the workflow for label 'e2e/run/gcp'...

File: go_lib/updater/updater.go
Comment (yalosev) [inline go_lib/updater/updater.go:496]: what it this input for?
I think, we have to describe it in the function description

File: deckhouse-controller/pkg/controller/deckhouse-release/updater/updater.go
Comment (ldmonster) [inline deckhouse-controller/pkg/controller/deckhouse-release/updater/updater.go:187]:
```suggestion
func (api *KubeAPI) IsKubernetesVersionAutomatic(ctx context.Context) error {
```
looks like it will be better without boolean return value. maybe return named error?

File: deckhouse-controller/pkg/controller/deckhouse-release/updater/updater.go
Comment (ldmonster) [inline deckhouse-controller/pkg/controller/deckhouse-release/updater/updater.go:204]: "Automatic" not const. we use it here only?
maybe it will be better to use constant value?

File: deckhouse-controller/pkg/controller/module-controllers/release/updater.go
Comment (ldmonster) [inline deckhouse-controller/pkg/controller/module-controllers/release/updater.go:223]:
```suggestion
		return fmt.Errorf("module deploy failed: %w", err)
```

File: deckhouse-controller/pkg/controller/deckhouse-release/updater/updater.go
Comment (ldmonster) [inline deckhouse-controller/pkg/controller/deckhouse-release/updater/updater.go:202]:
```suggestion
		return false, fmt.Errorf("check kubernetes version: failed to unmarshal cluster configuration: %w", err)
```

File: deckhouse-controller/pkg/controller/deckhouse-release/updater/updater.go
Comment (ldmonster) [inline deckhouse-controller/pkg/controller/deckhouse-release/updater/updater.go:191]:
```suggestion
		return false, fmt.Errorf("check kubernetes version: failed to get secret: %w", err)
```

File: go_lib/updater/updater.go
Comment (miklezzzz) [reply go_lib/updater/updater.go:496]: yeah, makes sense, the reason why I decided to add an argument is because almost all "metrics" magic takes place in the updater, but at the same time I don't want to update metrics for all current releases on each `PredictNextRelease` run, so I decided that having some sort of `run-designated` release would do the trick of updating only that release's metrics.
I find it to be a product of the updater's a bit strange design :)

File: deckhouse-controller/pkg/controller/module-controllers/release/updater.go
Comment (ldmonster) [inline deckhouse-controller/pkg/controller/module-controllers/release/updater.go:258]:
```suggestion
		return false, fmt.Errorf("check kubernetes version: failed to get secret: %w", err)
```

File: deckhouse-controller/pkg/controller/module-controllers/release/updater.go
Comment (ldmonster) [inline deckhouse-controller/pkg/controller/module-controllers/release/updater.go:269]:
```suggestion
		return false, fmt.Errorf("check kubernetes version: failed to unmarshal cluster configuration: %w", err)
```

File: go_lib/updater/updater.go
Comment (ldmonster) [inline go_lib/updater/updater.go:627]:
```suggestion
				if err := u.updateStatus(release, fmt.Sprintf("Awaiting for %s release to be deployed", previousPredictedRelease.GetName()), PhasePending); err != nil {
				u.logger.Error("update status", slog.String("error", err.Error()))
				}
```

**Comment by miklezzzz** [comment]
@ldmonster what's the point of suggestions without explanations? are you willing to do it every time a PR is reviewed? :))

File: go_lib/updater/updater.go
Comment (ldmonster) [reply go_lib/updater/updater.go:627]: why we ignore error?

File: deckhouse-controller/pkg/controller/module-controllers/release/updater.go
Comment (ldmonster) [reply deckhouse-controller/pkg/controller/module-controllers/release/updater.go:269]: why we not wrap error?

File: deckhouse-controller/pkg/controller/module-controllers/release/updater.go
Comment (ldmonster) [reply deckhouse-controller/pkg/controller/module-controllers/release/updater.go:258]: why we not wrap error?

File: deckhouse-controller/pkg/controller/deckhouse-release/updater/updater.go
Comment (ldmonster) [reply deckhouse-controller/pkg/controller/deckhouse-release/updater/updater.go:191]: why we not wrap error?

File: deckhouse-controller/pkg/controller/deckhouse-release/updater/updater.go
Comment (ldmonster) [reply deckhouse-controller/pkg/controller/deckhouse-release/updater/updater.go:202]: why we not wrap error?

File: deckhouse-controller/pkg/controller/module-controllers/release/updater.go
Comment (ldmonster) [reply deckhouse-controller/pkg/controller/module-controllers/release/updater.go:223]: why we not wrap error?

**Comment by ldmonster** [comment]
> @ldmonster what's the point of suggestions without explanations? are you willing to do it every time a PR is reviewed? :))

no problem.
i thought it's a known bad patterns to ignore errors and not wrap return errors

add comment to every not commented suggestions

File: go_lib/updater/updater.go
Comment (miklezzzz) [reply go_lib/updater/updater.go]: +

File: deckhouse-controller/pkg/controller/module-controllers/release/updater.go
Comment (miklezzzz) [reply deckhouse-controller/pkg/controller/module-controllers/release/updater.go]: fixed

File: deckhouse-controller/pkg/controller/module-controllers/release/updater.go
Comment (miklezzzz) [reply deckhouse-controller/pkg/controller/module-controllers/release/updater.go]: fixed

File: deckhouse-controller/pkg/controller/deckhouse-release/updater/updater.go
Comment (miklezzzz) [reply deckhouse-controller/pkg/controller/deckhouse-release/updater/updater.go]: fixed

File: deckhouse-controller/pkg/controller/deckhouse-release/updater/updater.go
Comment (miklezzzz) [reply deckhouse-controller/pkg/controller/deckhouse-release/updater/updater.go]: fixed

File: deckhouse-controller/pkg/controller/module-controllers/release/updater.go
Comment (miklezzzz) [reply deckhouse-controller/pkg/controller/module-controllers/release/updater.go]: fixed

File: deckhouse-controller/pkg/controller/deckhouse-release/updater/updater.go
Comment (miklezzzz) [reply deckhouse-controller/pkg/controller/deckhouse-release/updater/updater.go]: fixed, but it's going to take time and efforts to replace all ad-hoc strings in the project

**Review by ldmonster — APPROVED** [review]

**Review by yalosev — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @miklezzzz. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Pull request merged**
